openmoc/materialize.py

```diff
@@ -355,3 +360,314 @@ def load_openmc_mgxs_lib(mgxs_lib, geometry=None):
+
+    py_printf('NORMAL', 'Computing SPH factors...')
+
+    # Create an OpenMOC Geometry from the OpenCG Geometry
+    geometry = get_openmoc_geometry(mgxs_lib.opencg_geometry)
+
+    # Load the MGXS library data into the OpenMOC geometry
+    load_openmc_mgxs_lib(mgxs_lib, geometry)
+
+    # Initialize an OpenMOC TrackGenerator
+    geometry.initializeFlatSourceRegions()
+    track_generator = openmoc.TrackGenerator(geometry, num_azim, track_spacing)
+    track_generator.setZCoord(zcoord)
+    track_generator.generateTracks()
+
+    # Initialize an OpenMOC Solver
+    solver = openmoc.CPUSolver(track_generator)
+    solver.setConvergenceThreshold(fix_src_tol)
+    solver.setNumThreads(num_threads)
+    openmc_fluxes = _load_openmc_src(mgxs_lib, solver)
+
+    # Initialize SPH factors
+    num_groups = geometry.getNumEnergyGroups()
+    num_fsrs = geometry.getNumFSRs()
+
+    # Map FSRs to domains to compute domain-averaged fluxes
+    num_domains = len(mgxs_lib.domains)
+    fsrs_to_domains = np.zeros(num_fsrs)
+    for fsr in range(num_fsrs):
+        cell = geometry.findCellContainingFSR(fsr)
+
+        if mgxs_lib.domain_type == 'material':
+            fsrs_to_domains[fsr] = cell.getFillMaterial().getId()
+        else:
+            fsrs_to_domains[fsr] = cell.getId()
+
+    # Get all OpenMOC domains
+    if mgxs_lib.domain_type == 'material':
+        all_domains = geometry.getAllMaterials()
+    elif mgxs_lib.domain_type == 'cell':
+        all_domains = geometry.getAllMaterialCells()
+    else:
+        py_printf('ERROR', 'SPH factors cannot be applied for an OpenMC MGXS '
+                           'library of domain type %s', mgxs_lib.domain_type)
+
+    # Collect those domains for which we will apply SPH factors
+    if domains == 'all':
+        sph_domains = all_domains
+        sph_indices = np.arange(len(sph_domains))
+    elif domains == 'fissionable':
+        sph_domains = {}
+        sph_indices = []
+        for i, openmc_domain in enumerate(mgxs_lib.domains):
+            if all_domains[openmc_domain.id].isFissionable():
+                sph_domains[openmc_domain.id] = all_domains[openmc_domain.id]
+                sph_indices.append(i)
+    else:
+        py_printf('ERROR', 'SPH factors cannot be applied for domain type '
+                           '%s which is not supported', str(domains))
+
+    # Initialize array of domain-averaged fluxes and SPH factors
+    openmoc_fluxes = np.zeros((num_domains, num_groups))
+
+    # Outer SPH loop over domains of interest
+    for i in range(max_domain_iters):
+        for domain_id in sph_domains:
+            py_printf('NORMAL', 'SPH outer iteration %d: %s %d',
+                                i, mgxs_lib.domain_type.capitalize(), domain_id)
+
+            inner_sph = np.ones((num_domains, num_groups))
+
+            for j in range(max_fix_src_iters):
+
+                # Run fixed source calculation with throttled verbosity
```

Comment:
The reason is that the output is very long and difficult to comprehend if it is not throttled. The `Solver::computeFlux()` fixed source calculation scheme will report its own residuals after each transport sweep. Combine this with the residuals from the inner / outer loops for the SPH calculation and there is a lot going on in the output.

To compromise on this, I added a new `throttle_output` parameter to this routine which is `True` by default. In this case, it will use verbosity throttling to suppress the output from the fixed source calculations, and restore the original verbosity level after the completion of the SPH calculation.